dev/AppNotifications/AppNotificationTelemetry.h

```diff
@@ -0,0 +1,318 @@
+        winrt::hresult hr,
+        std::wstring const& appId) noexcept try
+    {
+        if (c_maxEventLimit >= UpdateLogEventCount())
+        {
+            TraceLoggingClassWriteMeasure(
+                "RemoveAllAsyncByAPI",
+                TelemetryPrivacyDataTag(PDT_ProductAndServicePerformance),
+                _GENERIC_PARTB_FIELDS_ENABLED,
+                TraceLoggingHexUInt32(hr, "OperationResult"),
+                TraceLoggingWideString(appId.c_str(), "AppId"),
+                TraceLoggingBool(IsPackagedApp(), "IsAppPackaged"),
+                TraceLoggingWideString(GetAppName(), "AppName"));
+        }
+    }
+    CATCH_LOG()
+
+    DEFINE_EVENT_METHOD(GetAllAsyncByAPI)(
+        winrt::hresult hr,
+        std::wstring const& appId) noexcept try
+    {
+        if (c_maxEventLimit >= UpdateLogEventCount())
+        {
+            TraceLoggingClassWriteMeasure(
+                "GetAllAsyncByAPI",
+                TelemetryPrivacyDataTag(PDT_ProductAndServicePerformance),
+                _GENERIC_PARTB_FIELDS_ENABLED,
+                TraceLoggingHexUInt32(hr, "OperationResult"),
+                TraceLoggingWideString(appId.c_str(), "AppId"),
+                TraceLoggingBool(IsPackagedApp(), "IsAppPackaged"),
+                TraceLoggingWideString(GetAppName(), "AppName"));
+        }
+    }
+    CATCH_LOG()
+
+private:
+    wil::srwlock m_lock;
+    ULONGLONG m_lastFiredTick = 0;
+    UINT m_eventCount = 0;
+
+    static constexpr ULONGLONG c_logPeriod = 1000; // One second
+    static constexpr UINT c_maxEventLimit = 10;
+
+    UINT UpdateLogEventCount() noexcept
+    {
+        ULONGLONG currentTick = GetTickCount64();
+
+        auto lock{ m_lock.lock_exclusive() };
+
+        // Only fire limiting events every log period to prevent too many events from being fired
+        if ((currentTick - m_lastFiredTick) > c_logPeriod)
+        {
+            m_eventCount = 0;
+            m_lastFiredTick = currentTick;
+        }
+
+        m_eventCount++;
+
+        return m_eventCount;
+    }
+
+    inline bool IsPackagedApp()
+    {
+        static const bool isPackagedApp = AppModel::Identity::IsPackagedProcess();
+
+        return isPackagedApp;
+    }
+
+    inline const wchar_t* GetAppName()
+    {
+        static const std::wstring appName = IsPackagedApp() ? GetAppNamePackaged() : GetAppNameUnpackaged();
+
+        return appName.c_str();
+    }
+
+    std::wstring GetAppNamePackaged() noexcept
+    {
+        wchar_t appUserModelId[APPLICATION_USER_MODEL_ID_MAX_LENGTH]{};
+
+        UINT32 appUserModelIdSize = ARRAYSIZE(appUserModelId);
+        auto result = GetCurrentApplicationUserModelId(&appUserModelIdSize, appUserModelId);
+        if (result != ERROR_SUCCESS)
+        {
+            wcscpy_s(appUserModelId, L"AppUserModelId not found");
+            LOG_WIN32(result);
+        }
+
+        return appUserModelId;
+    }
+
+    PCWSTR CensorFilePath(PCWSTR path) noexcept
+    {
+        if (path)
```

Comment:
If Path is empty or null, this should throw.

Reply:
not a good idea because telemetry events are all no throw, so they can be used in situation where exceptions are already in flight or in destructors.

This is a private function, even if it throws, I'll have to catch it before it exits the telemetry log function and there's nothing more that I'll be able to do at that point.

The goal here is to do a best effort of logging as mush details as possible but nothing more.

Reply:
not a good idea because telemetry events are all no throw, so they can be used in situation where exceptions are already in flight or in destructors.

This is a private function, even if it throws, I'll have to catch it before it exits the telemetry log function and there's nothing more that I'll be able to do at that point.

The goal here is to do a best effort of logging as mush details as possible but nothing more.

Reply:
Best effort is not an option at this layer. The Logging wrappers do the best effort like LogRegister() which is wrapped in a try catch_log block. So we would never throw at the SDK layer even if something bad happens during logging.
The helpers that it uses to achieve its goal should throw. Otherwise, we end up with non-determinist behavior in the Logging wrapper layers.